library/src/main/java/com/bumptech/glide/manager/DefaultConnectivityMonitor.java

```diff
@@ -27,43 +31,97 @@
 
   private boolean isRegistered;
 
-  private final BroadcastReceiver connectivityReceiver =
-      new BroadcastReceiver() {
-        @Override
-        public void onReceive(@NonNull Context context, Intent intent) {
-          boolean wasConnected = isConnected;
-          isConnected = isConnected(context);
-          if (wasConnected != isConnected) {
-            if (Log.isLoggable(TAG, Log.DEBUG)) {
-              Log.d(TAG, "connectivity changed, isConnected: " + isConnected);
-            }
-
-            listener.onConnectivityChanged(isConnected);
-          }
-        }
-      };
+  private final BroadcastReceiver connectivityReceiver;
+  private final ConnectivityManager.NetworkCallback callback;
 
   DefaultConnectivityMonitor(@NonNull Context context, @NonNull ConnectivityListener listener) {
     this.context = context.getApplicationContext();
     this.listener = listener;
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      connectivityReceiver = null;
+      callback =
+          new ConnectivityManager.NetworkCallback() {
+            @Override
+            public void onAvailable(@NonNull Network network) {}
+
+            @Override
+            public void onLost(@NonNull Network network) {
+              onConnectivityChanged(false);
+            }
+
+            @Override
+            public void onCapabilitiesChanged(
+                @NonNull Network network, @NonNull NetworkCapabilities networkCapabilities) {
+              boolean connected =
+                  networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                      && networkCapabilities.hasCapability(
+                          NetworkCapabilities.NET_CAPABILITY_VALIDATED);
+              onConnectivityChanged(connected);
```

Comment:
Can this just call isConnected? It seems like you've updated that method anyway?

Reply:
ConnectivityManager.getNetworkCapabilities is synchronous method as this is prone to race conditions